### ISSUE TITLE
run-with-sudo function appears to be broken on macOS 10.14

### ISSUE THREAD

**Issue by jakewilliami** [0zz4r/bashflix#8]
After it waits for my password the first time, every other time, in quick succession, it will ask for password, and immediately tell me it is incorrect.  Then it gives up.

P.S.: nice work on this!  This is really neat.

```
Password is mandatory!

notes: the password will be used to install third-party software (run apt and pip3 with sudo)

Insert your password:
Checking OS ...
Checking bashflix requirements ...
Looking for Homebrew ...
Preparing to install PIP3 ...
Preparing to install MPV ...
Already up-to-date.
Error: node 14.0.0 is already installed
To upgrade to 14.1.0, run `brew upgrade node`.
Warning: python 3.7.7 is already installed and up-to-date
To reinstall 3.7.7, run `brew reinstall python`
Warning: mpv 0.32.0_1 is already installed and up-to-date
To reinstall 0.32.0_1, run `brew reinstall mpv`
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
./install.sh: line 81: cd: rarbgapi: No such file or directory
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
```

**Comment by astavares** [comment]
try to run it again, it will probably work. I will fix the issue nevertheless

**Comment by jakewilliami** [comment]
Unfortunately it still isn't working.

**Comment by astavares** [comment]
new commit. try again please

**Comment by jakewilliami** [comment]
Hmm, still not working.  This time it does indeed install everything it needs to, but then it fails again:
```
[jakeireland@jake-mbp2017: bashflix] [master] $ ./install.sh
Password is mandatory!

notes: the password will be used to install third-party software (run apt and pip3 with sudo)

Insert your password:
Looking for Homebrew ...
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 3 taps (homebrew/cask-versions, homebrew/core and homebrew/cask).
==> New Formulae
dmagnetic                        semgrep                          sollya                           xxh
==> Updated Formulae
imagemagick ✔     awscli@1          couchdb           helmfile          mg                opencc            terragrunt
libressl ✔        bde               duplicity         hlint             movgrab           picat             testssl
rust ✔            bison             e2fsprogs         imagemagick@6     mu                plantuml          tmux-xpanes
abcmidi           bmake             exploitdb         jupyterlab        mvtools           pngpaste          whistle
aliyun-cli        citus             ext2fuse          katago            nanopb-generator  poetry            wildfly-as
ammonite-repl     clojure-lsp       frugal            kubeseal          ngt               protobuf          wtfutil
amqp-cpp          clojurescript     gitleaks          leakcanary-shark  now-cli           protoc-gen-go     z3
appium            closure-compiler  gnumeric          libfabric         oauth2_proxy      ripgrep           zero-install
ask-cli           clp               goffice           libplctag         oha               serf
aws-sdk-cpp       consul            gopass            maxwell           okteto            tealdeer
==> Updated Casks
115browser                diskcatalogmaker          ipe                       plexamp                   teamdrive
abstract                  dotnet-sdk-preview        ipepresenter              postman                   telegram-desktop
alt-tab                   empoche                   maintenance               publish-or-perish         telegram-desktop-dev
anki                      exodus                    middleclick               raven-reader              trezor-bridge
appzapper                 freeyourmusic             mochi                     resilio-sync              vagrant-vmware-utility
audirvana                 fsnotes                   onyx                      retroarch-metal           vysor
busycal                   gemini                    openaudible               slack-beta                xctu
busycontacts              geogebra                  pdf-converter-master      spires                    youdaodict
deeper                    gloomhaven-helper         pdf-toolbox               stack-stack
==> Deleted Casks
glip

==> Upgrading 1 outdated package:
node 14.0.0 -> 14.2.0
==> Upgrading node 14.0.0 -> 14.2.0
==> Downloading https://homebrew.bintray.com/bottles/node-14.2.0.mojave.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/70/7024e321dc5e6ca079ad80f44eb4115e729d9809a87360fe6c727a62ba7c4491?__gda__=exp=15
######################################################################## 100.0%
==> Pouring node-14.2.0.mojave.bottle.tar.gz
==> Caveats
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d
==> Summary
🍺  /usr/local/Cellar/node/14.2.0: 4,659 files, 60.7MB
Removing: /usr/local/Cellar/node/14.0.0... (4,659 files, 60.7MB)
==> Checking for dependents of upgraded formulae...
==> No dependents found!
Warning: node 14.2.0 is already installed and up-to-date
To reinstall 14.2.0, run `brew reinstall node`
Warning: python3 3.7.7 already installed
Warning: python 3.7.7 already installed
Warning: python 3.7.7 is already installed and up-to-date
To reinstall 3.7.7, run `brew reinstall python`
Already up-to-date.
Warning: mpv 0.32.0_1 already installed
Updating Homebrew...
Warning: mpv 0.32.0_1 is already installed and up-to-date
To reinstall 0.32.0_1, run `brew reinstall mpv`
Updating Homebrew...
==> Downloading https://get.videolan.org/vlc/3.0.10/macosx/vlc-3.0.10.dmg
==> Downloading from https://videolan.mirrors.theom.nz/vlc/3.0.10/macosx/vlc-3.0.10.dmg
######################################################################## 100.0%
==> Verifying SHA-256 checksum for Cask 'vlc'.
==> Installing Cask vlc
==> Moving App 'VLC.app' to '/Applications/VLC.app'.
==> Linking Binary 'vlc.wrapper.sh' to '/usr/local/bin/vlc'.
🍺  vlc was successfully installed!
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
./install.sh: line 65: cd: rarbgapi: No such file or directory
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
Password:Sorry, try again.
Password:
sudo: no password was provided
sudo: 1 incorrect password attempt
```

I do have a recommendation as well (while you're editing the install script).  I get a lot of "Warning: x is already installed and up-to-date" messages.  Perhaps you could check they are installed before trying to install them?:
```
if ! grep "^${package}$" $(brew list) > /dev/null 2>&1; then
    [install packages]
fi
```
However, this isn't detrimental to the script; only a suggestion.  Great work!

**Comment by astavares** [comment]
try run this
`bash <(curl https://raw.githubusercontent.com/astavares/bashflix/master/install.sh)`
I have a mac and I don't get that output
are you sure the password is correct? 
thank you very much for the suggestion, I will add it!

**Comment by jakewilliami** [comment]
That command doesn't work either.
I figured it's be really difficult for you to fix bugs when you can't actually recreate the problem, so I had a play around.  I found that, 


Another option would be to simply put

**Comment by jakewilliami** [comment]
### That command doesn't work either...
I figured it's be really difficult for you to fix bugs when you can't actually recreate the problem, so I've been playing around with it for hours and I just can't get it to work using the `run-with-sudo` function, but I'm unsure why it isn't working.  I am using 
```bash
jakeireland@jake-mbp2017: bashflix] [master] $ bash --version
GNU bash, version 5.0.17(1)-release (x86_64-apple-darwin18.7.0)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```
and
```bash
jakeireland@jake-mbp2017: bashflix] [master] $ sudo --version
Sudo version 1.8.29
Sudoers policy plugin version 1.8.29
Sudoers file grammar version 46
Sudoers I/O plugin version 1.8.29
```
If you insist on using the `run-with-sudo` function, I recommend that, rather than `echo "${password}" | sudo -S ${@}`, you use
```bash
/usr/bin/sudo -S bash -c "exec < /dev/tty; ${@}" <<< "${password}"
```
to avoid [any future `sudo` issues](https://askubuntu.com/questions/1121283/bash-sudo-s-mypassword-interactive-shell-issue/1121309#1121309).

On that note, if you do prefer to use `run-with-sudo`, then just to make it a bit cleaner, you could consider adding your prompt to the read function:
```bash
read -sp $'Password is mandatory!
\n\nNotes: the password will be used to install third-party software (run apt and pip3 with sudo)
\n\nInsert your password: ' password
```

---

### My solution:
I think **the option I prefer** (and one that will replace the `run-with-sudo` function) is to, at the start of the script, check whether or not the user is running the script as `sudo`; if not, ensure that it is
```bash
case $EUID in
   0) : ;; # we are running script as root so we are okay
   *)  /usr/bin/sudo $0 "${@}" ;; # not root, become root for the rest of this session (and ask for the sudo password only once)
esac
```
Then all instances of `run-with-sudo` can be replaced simply with `sudo`, and the `run-with-sudo` function can be removed.

---

### Other points:
I have another few things that need to be changed after having a look at the script:
- The Homebrew install URL should end with a `.sh` (see [the `.sh` link](https://raw.githubusercontent.com/Homebrew/install/master/install.sh) compared to [the one in your script](https://raw.githubusercontent.com/Homebrew/install/master/instal))
- `brew install git` is not needed anymore (see lines 156--159 in [the Homebrew install script](https://github.com/Homebrew/install/blob/master/install.sh))
- I notice we are installing `webtorrent-cli` using `node`; this isn't necessarily a problem, but you can now obtain it via `brew` (`brew install webtorrent-cli`).  I'm not sure which is preferable to you 😄 
- Finally, perhaps a better thing to do than multiple `if` statements checking for the system running the script and installing dependencies from there is to do something like this:
    ```bash
    if [[ $(uname -s) == "Darwin" ]]
    then
        KERNEL="Darwin"
        OS="macos"
    elif [[ $(uname -s) == "Linux" ]]
    then
        KERNEL="Linux"
        if [[ -f /etc/arch-release ]]
        then
            OS="arch"
        elif [[ -f /etc/debian_version ]]
        then
            OS="debian"
        elif [[ -f /etc/redhat-release ]]
        then
            OS="fedora"
        fi
    else
        exit 1
    fi

    if command -v brew > /dev/null 2>&1 # check if brew is installed
    then
        PACMAN='brew install'
        PACSEARCH='brew list'
    else # if not brew, check for OS
        declare -A osInfo;
        osInfo[/etc/redhat-release]='sudo dnf --assumeyes install'
        osInfo[/etc/arch-release]='sudo pacman -S --noconfirm'
        osInfo[/etc/gentoo-release]='sudo emerge'
        osInfo[/etc/SuSE-release]='sudo zypper in'
        osInfo[/etc/debian_version]='sudo apt-get --assume-yes install'
    
        declare -A osSearch;
        osSearch[/etc/redhat-release]='dnf list installed'
        osSearch[/etc/arch-release]='pacman -Qq'
        osSearch[/etc/gentoo-release]="cd /var/db/pkg/ && ls -d */*| sed 's/\/$//'"
        osSearch[/etc/SuSE-release]='rpm -qa'
        osSearch[/etc/debian_version]='dpkg -l' # previously `apt list --installed`.  Can use `sudo apt-cache search`.
    
        for f in "${!osInfo[@]}"
        do
            if [[ -f $f ]];then
                PACMAN="${osInfo[$f]}"
            fi
        done
    
        for s in "${!osSearch[@]}"
        do
            if [[ -f $s ]];then
                PACSEARCH="${osSearch[$s]}"
            fi
        done
    fi
    ```
    Now we are able to easily install a formula or library by defining some functions checking if the user already has it installed:
    ```bash
    formula_install() {
        PACINSTALL=false
        for i in "${@}"; do
            if ! command -v "${i}" > /dev/null 2>&1
            then
                echo -e "Installing ${i}"
                $PACMAN "${i}" && PACINSTALL=true
            fi
        done
        if [ "${PACINSTALL}" == true ]
        then
            return 0
        else
            return 1
        fi
    }
    
    library_install() {
        PACINSTALL=false
        for i in "${@}"; do
            if ! $PACSEARCH | grep "^${i}$" > /dev/null 2>&1
            then
                echo "Installing ${i}"
                $PACMAN "${i}" && PACINSTALL=true
            fi
        done
        if [[ "${PACINSTALL}" == true ]]
        then
            return 0
        else
            return 1
        fi
    }
    ```
    We also need to define the formula or library:
    ```bash
    #define the formula that the majority of OSs use
    #so that we only have to redefine formula minimally, as required
    PYTHON3='python3'
    PIP3='python3-pip'
    
    case $OS in
        macos)
            PYTHON3='python' # Includes pip3 on macOS
            ;;
        arch)
            PYTHON3='python'
            PIP3='python-pip' 
            ;;
        debian) ;;
        fedora) ;;
    esac
    ```
    And now we can easily get the formula or libraries we need:
    ```bash
    formula_install "${PYTHON3}"
    library_install "${PIP3}"
    ```
    Of course, this may all be way overkill for a simple install script...it's your decision! 😆 All the best!

**Comment by 0zz4r** [comment]
Hi. Just an update about this. I've created a dev version of the installer using your code that I will merge to master after I (or someone else) properly test it. You can already use it though:
bash <(curl https://raw.githubusercontent.com/astavares/bashflix/dev/install.sh)
Thanks very much for the suggestion and the code!

**Comment by steeph-k** [comment]
I looked at the script before trying it and was about to suggest the same thing for different reasons. Great to see it's already done. It's working nicely. Thanks for the work!

**Comment by 0zz4r** [comment]
I've implemented part of your code in the current version and this bug is fixed.
